File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.cpp
class Solution {
public:
    int myAtoi(string s) {
        int n = (int)s.size();
        int i = 0;
        // skip whitespace
        while (i < n && s[i] == ' ') i++;
        // account for sign
        int sign = 1;
        if (i < n && (s[i] == '-' || s[i] == '+')){
            if (s[i] == '-')
                sign = -1;
            i++;
        }
        // process digits, clamping before overflow
        int ans = 0;
        while (i < n && isdigit(s[i])){
            int digit = s[i]-'0';
            if (ans > (INT_MAX - digit) / 10)
                return sign == 1 ? INT_MAX : INT_MIN;
            ans = ans * 10 + digit;
            i += 1;
        }
        return sign * ans;
    }
};

File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.py
class Solution:
    def myAtoi(self, s: str) -> int:
        if s == "":
            return 0
        i = 0
        n = len(s)
        while i < n and s[i] == ' ':
            i += 1
        if i == n: return 0
        sign = 1
        if s[i] == '-' or s[i] == '+':
            if s[i] == '-':
                sign *= -1
            i += 1
        MAX = 2**31 - 1
        MIN = -2**31
        ans = 0
        while i < n and s[i].isdigit():
            digit = int(s[i])
            if ans > (MAX - digit) / 10:
                print('dfafdsaf')
                return MAX if sign == 1 else MIN
            ans = ans * 10 + digit
            i += 1
        return sign * ans